### ISSUE TITLE
feature request: Dynamic SSO Identity Provider selection mechanism (beyond email domain)

### ISSUE THREAD

**Issue by cesdperez** [logto-io/logto#5662]
## Summary

We're looking into using logto.io for SSO across our B2B products, but we've hit a snag. The current IdP selection relies mainly on email domains, which doesn't work well for us because:

- Some of our clients use generic email providers like Gmail.
- Others have multiple IdPs for the same domain.

We need a more flexible way to route authentication requests.

## Proposed Solution

Allow applications to pass extra parameters (see for example Keycloak's `idpHint` parameter) to help select the right IdP.
Alternatively, add customizable rules that can look at things like user input, IP address, or metadata to dynamically pick the correct IdP.

## Benefits

- Makes logto.io work for more types of organizations and use cases.
- Gives users a smoother authentication experience.
- Keeps logto.io competitive with other SSO solutions.

We think a more flexible IdP selection feature would be a big help for users and companies with complex authentication setups. We'd love to hear any ideas or suggestions from the community or the logto.io team on how to tackle this.

**Comment by simeng-li** [comment]
Hi @cesdperez , thanks for your feedback. 

We actually have a direct sign-in feature that allows users to jump directly to a specific SSO IdP's sign-in page, without needing a valid email domain. Would this address your issue?

However, this feature requires you to have a particular SSO IdP pre-configured. For multiple SSO IdP enabled, we still need to use user email to identify the enabled SSO connectors.

**Comment by simeng-li** [comment]
Glad to hear. It's a new feature we just added. We are working on the documentation. I'll keep you updated once it's released.

**Comment by simeng-li** [comment]
@cesdperez please check this [doc](https://docs.logto.io/docs/recipes/openid-connect/authentication-parameters/). Let us know if this helps.

**Comment by cesdperez** [comment]
Yes, that would be what I'm looking for 👍🏼. I haven't verified the functionality, tho. Thanks for the quick response.

Nonetheless, my use case requires that for any kind of enterprise SSO connection (either OIDC or SAML).

Any plans to bring that to SAML too?

**Comment by cesdperez** [comment]
I see that the email domain is still a required field when creating an SSO Enterprise connector 
<img width="856" alt="image" src="https://github.com/logto-io/logto/assets/16674679/5160521c-16ef-4d97-b7e3-d9510787fc34">.

For my use case, I'd need this to be optional.

I'm using version 1.15.

**Comment by simeng-li** [comment]
For direct sign-in, you may leave this field empty, it should jump the email domain validation step.  Let us know if you met any blockers.

**Comment by cesdperez** [comment]
I'm not being able to make this work.

I have a working enterprise SSO connector, and I'm using the React SDK like this:

```
import { useLogto } from '@logto/react';
import { useState, useEffect } from 'react';

const Home = () => {
    const { isAuthenticated, getIdTokenClaims, signIn, signOut } = useLogto();
    const [userId, setUserId] = useState('');

    useEffect(() => {
        (async () => {
            if (isAuthenticated) {
                const claims = await getIdTokenClaims();
                setUserId(claims.sub);
            }
        })();
    }, [isAuthenticated]);

    return (
        <div>
            {userId && <p>Logged in as {userId}</p>}
            {isAuthenticated ? (
                <button onClick={() => signOut('http://localhost:5173')}>
                    Sign Out
                </button>
            ) : (
                <button onClick={() => signIn({
                    redirectUri: 'http://localhost:5173/callback',
                    directSignIn: 'sso:ac6lupl1o2a6'
                })}>
                    Sign In
                </button>
            )}
        </div>
    );
};

export default Home;
```

When I click "sign in" I see the default Logto login screen (asking for username and password). I see this in the browser console which doesn't look correct:
```
Navigated to https://3001-logtoio-demo-cgr8sxlpiwa.ws-eu114.gitpod.io/direct/undefined/undefined?fallback=sign-in
Navigated to https://3001-logtoio-demo-cgr8sxlpiwa.ws-eu114.gitpod.io/sign-in
...
```

Enterpise SSO is enabled in my "Sign-in experience" configuration. The connector id `sso:ac6lupl1o2a6` is correct. Not sure what else I could be missing.

**Comment by simeng-li** [comment]
Apologies for missing your latest response.

Please check whether SSO is enabled in your sign-in experience settings. We've updated our documentation since the last reply — feel free to review the latest version and let us know if you're still experiencing issues. Close the issue for now, reopen it  is needed.
